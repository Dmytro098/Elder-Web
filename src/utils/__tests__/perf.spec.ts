import normalizeSnapshot from '../normalizeSnapshot';

class PerformanceObserverMock {
  cb: (any) => void;

  constructor(cb) {
    this.cb = cb;
  }

  observe() {
    this.cb({ getEntries: () => [{ name: 'Page-xxxxxxxx', duration: 0.05 }] });
  }

  disconnect() {
    this.cb = null;
  }
}

describe('#perf', () => {
  it('works in performance', () => {
    function MockPage() {
      this.uid = 'xxxxxxxx';
      this.htmlString = '';
      this.settings = {
        debug: {
          performance: true,
        },
      };
    }
    const calls = [];
    jest.mock('perf_hooks', () => ({
      PerformanceObserver: PerformanceObserverMock,
      performance: {
        mark: (i) => calls.push(`mark ${i}`),
        measure: (i) => calls.push(`measure ${i}`),
        clearMarks: (i) => calls.push(`clearMarks ${i}`),
      },
    }));
    const mockPage = new MockPage();
    // eslint-disable-next-line global-require
    const perf = require('../perf').default;

    // mutate
    perf(mockPage);

    mockPage.perf.start('test');
    mockPage.perf.end('test');
    mockPage.perf.stop();

    expect(normalizeSnapshot(mockPage)).toMatchSnapshot();
    expect(calls).toEqual([
      'clearMarks Page-xxxxxxxx',
      'mark test-start-xxxxxxxx',
      'mark test-end-xxxxxxxx',
      'measure test-xxxxxxxx',
    ]);
  });
  it('works in non performance', () => {
    function MockPage() {
      this.uid = 'xxxxxxxx';
      this.htmlString = '';
      this.settings = {
        debug: {
          performance: false,
        },
      };
    }
    const calls = [];
    jest.mock('perf_hooks', () => ({
      PerformanceObserver: PerformanceObserverMock,
      performance: {
        mark: (i) => calls.push(`mark ${i}`),
        measure: (i) => calls.push(`measure ${i}`),
        clearMarks: (i) => calls.push(`clearMarks ${i}`),
      },
    }));
    const mockPage = new MockPage();
    // eslint-disable-next-line global-require
    const perf = require('../perf').default;

    // mutate
    perf(mockPage);

    mockPage.perf.start('test');
    mockPage.perf.end('test');
    mockPage.perf.stop();

    expect(calls).toEqual([]);
  });
});
