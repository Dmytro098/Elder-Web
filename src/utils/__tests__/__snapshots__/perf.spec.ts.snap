// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#perf works in performance 1`] = `
Object {
  "htmlString": "",
  "perf": Object {
    "end": [Function],
    "start": [Function],
    "stop": [Function],
    "timings": Array [
      Object {
        "duration": 0.05,
        "name": "Page",
      },
    ],
  },
  "settings": Object {
    "debug": Object {
      "performance": true,
    },
  },
  "uid": "xxxxxxxx",
}
`;
